lib/FrontendTool/FrontendTool.cpp

```diff
@@ -739,38 +743,40 @@ static Optional<bool> dumpASTIfNeeded(CompilerInvocation &Invocation,
 static void emitReferenceDependenciesForAllPrimaryInputsIfNeeded(
     CompilerInvocation &Invocation, CompilerInstance &Instance) {
   if (Invocation.getFrontendOptions()
-          .InputsAndOutputs.supplementaryOutputs()
-          .ReferenceDependenciesFilePath.empty())
-    return;
-  if (Instance.getPrimarySourceFiles().empty()) {
+          .InputsAndOutputs.hasReferenceDependenciesPath() &&
+      Instance.getPrimarySourceFiles().empty()) {
     Instance.getASTContext().Diags.diagnose(
         SourceLoc(), diag::emit_reference_dependencies_without_primary_file);
     return;
   }
   for (auto *SF : Instance.getPrimarySourceFiles()) {
-    std::string referenceDependenciesFilePath =
-        Invocation.getPrimarySpecificPathsForSourceFile(*SF)
-            .SupplementaryOutputs.ReferenceDependenciesFilePath;
-    emitReferenceDependenciesIfNeeded(Instance.getASTContext().Diags, SF,
+    const std::string &referenceDependenciesFilePath =
+        Invocation.getReferenceDependenciesFilePathForPrimary(
+            SF->getFilename());
+    if (!referenceDependenciesFilePath.empty())
+      (void)emitReferenceDependencies(Instance.getASTContext().Diags, SF,
                                       *Instance.getDependencyTracker(),
                                       referenceDependenciesFilePath);
   }
 }
 
 static bool writeTBDIfNeeded(CompilerInvocation &Invocation,
                              CompilerInstance &Instance) {
-  StringRef TBDPath = Invocation.getFrontendOptions()
-                          .InputsAndOutputs.supplementaryOutputs()
-                          .TBDPath;
-  if (TBDPath.empty())
-    return false;
-  auto installName = Invocation.getFrontendOptions().TBDInstallName.empty()
-                         ? "lib" + Invocation.getModuleName().str() + ".dylib"
-                         : Invocation.getFrontendOptions().TBDInstallName;
-
-  return writeTBD(Instance.getMainModule(),
-                  Invocation.getSILOptions().hasMultipleIGMs(), TBDPath,
-                  installName);
+  return Invocation.getFrontendOptions()
+      .InputsAndOutputs.forEachInputProducingSupplementaryOutput(
+          [&](const InputFile &input) -> bool {
+            const std::string &TBDPath = input.TBDPath();
+            if (TBDPath.empty())
+              return false;
+            auto installName =
+                Invocation.getFrontendOptions().TBDInstallName.empty()
+                    ? "lib" + Invocation.getModuleName().str() + ".dylib"
+                    : Invocation.getFrontendOptions().TBDInstallName;
+
+            return writeTBD(Instance.getMainModule(),
```

Comment:
Just to confirm, are you pointing out that the `TBDPath` does *not* belong in `PrimarySpecificPaths`? If that's the case (and I suspect it may be), I'll move it out. Thanks.

Reply:
That would be one reasonable resolution here. Another would be to remove the loop and have a getTBDPathForWholeModule accessor like the AtMostOnePrimary ones you already have, which just looks at the first input file all the time. I don't think I have a preference between those two.

Reply:
I've now pushed a redone `writeTBDIfNeeded`. If will fail an assertion if a TBDPath is specified when not in WMO mode. Is that the right behavior?